File: client/src/services/contactService.js
import axios from 'axios';
import { handleApiError } from '../utils/errorHandler';

const API_URL = import.meta.env.VITE_API_URL;

const contactService = {
    getContacts: async (orderBy = 'name', orderType = 'asc', page = 1, perPage = 50, filters = {}) => {
        try {
            const params = {
                order_by: orderBy,
                order_type: orderType,
                page: page,
                per_page: perPage,
                ...filters
            };

            const response = await axios.get(`${API_URL}/api/contacts`, {
                params: params,
                withCredentials: true
            });
            return response.data;
        } catch (error) {
            console.error("Error fetching contacts:", error);
            throw handleApiError(error);
        }
    },

    getContact: async (id) => {
        try {
            const response = await axios.get(`${API_URL}/api/contacts/${id}`, { withCredentials: true });
            return response.data;
        } catch (error) {
            console.error("Error fetching contact:", error);
            throw handleApiError(error);
        }
    },

    createContact: async (contactData, avatar = null) => {
        try {
            const formData = new FormData();

            Object.entries(contactData).forEach(([key, value]) => {
                if (value !== null && value !== undefined) {
                    if (Array.isArray(value)) {
                        if (value.length > 0) {
                            formData.append(`contact[${key}]`, JSON.stringify(value));
                        }
                    }
                    else if (typeof value === 'object' && !Array.isArray(value)) {
                        formData.append(`contact[${key}]`, JSON.stringify(value));
                    }
                    else if (typeof value === 'string') {
                        if (value.trim() !== '') {
                            formData.append(`contact[${key}]`, value);
                        }
                    }
                    else {
                        formData.append(`contact[${key}]`, value);
                    }
                }
            });

            if (avatar) {
                formData.append('avatar', avatar);
            }

            const response = await axios.post(`${API_URL}/api/contacts`, formData, {
                withCredentials: true,
                headers: {
                    'Content-Type': 'multipart/form-data'
                }
            });

            return response.data;
        } catch (error) {
            if (error.response && error.response.status === 422) {
                console.error("Validation error:", error.response.data);
                throw new Error(JSON.stringify(error.response.data));
            }
            console.error("Error creating contact:", error);
            throw handleApiError(error);
        }
    },

    updateContact: async (id, contactData, avatar = null) => {
        try {
            const formData = new FormData();

            Object.entries(contactData).forEach(([key, value]) => {
                if (value !== null && value !== undefined) {
                    if (Array.isArray(value)) {
                        if (value.length > 0) {
                            formData.append(`contact[${key}]`, JSON.stringify(value));
                        }
                    }
                    else if (typeof value === 'object' && !Array.isArray(value)) {
                        formData.append(`contact[${key}]`, JSON.stringify(value));
                    }
                    else if (typeof value === 'string') {
                        if (value.trim() !== '') {
                            formData.append(`contact[${key}]`, value);
                        }
                    }
                    else {
                        formData.append(`contact[${key}]`, value);
                    }
                }
            });

            if (avatar) {
                formData.append('avatar', avatar);
            }

            const response = await axios.put(`${API_URL}/api/contacts/${id}`, formData, {
                withCredentials: true,
                headers: {
                    'Content-Type': 'multipart/form-data'
                }
            });

            return response.data;
        } catch (error) {
            if (error.response && error.response.status === 422) {
                console.error("Validation error:", error.response.data);
                throw new Error(JSON.stringify(error.response.data));
            }
            console.error("Error updating contact:", error);
            throw handleApiError(error);
        }
    },

    deleteContact: async (id) => {
        try {
            await axios.delete(`${API_URL}/api/contacts/${id}`, { withCredentials: true });
            return { success: true };
        } catch (error) {
            console.error("Error deleting contact:", error);
            throw handleApiError(error);
        }
    },

    getContactFields: async () => {
        try {
            const response = await axios.get(`${API_URL}/api/contacts/fields`, { withCredentials: true });
            return response.data.fields;
        } catch (error) {
            console.error("Error fetching contact fields:", error);
            throw handleApiError(error);
        }
    },

    mergeContacts: async (primaryId, secondaryIds, contactData = {}) => {
        console.log('Primary ID:', primaryId);
        console.log('Secondary IDs:', secondaryIds);
        console.log('Contact data:', contactData);
        try {
            const payload = {
                primary_contact_id: primaryId,
                secondary_contact_ids: secondaryIds
            };

            if (Object.keys(contactData).length > 0) {
                payload.contact = contactData;
            }

            const response = await axios.post(`${API_URL}/api/contacts/merge`, payload, {
            withCredentials: true
            });
            return response.data;
        } catch (error) {
            console.error("Error merging contacts:", error);
            throw handleApiError(error);
        }
    },

    getContactCount: async () => {
        try {
            const response = await axios.get(`${API_URL}/api/contacts/count`, { withCredentials: true });
            return response.data;
        } catch (error) {
            console.error("Error fetching contact count:", error);
            throw handleApiError(error);
        }
    },

    getCompanies: async () => {
        try {
            const response = await axios.get(`${API_URL}/api/contacts/companies`, { withCredentials: true });
            return response.data;
        } catch (error) {
            console.error("Error fetching companies:", error);
            throw handleApiError(error);
        }
    },

    exportContacts: async () => {
        try {
            const response = await axios.get(`${API_URL}/api/contacts/export`, {
                withCredentials: true,
                responseType: 'blob'
            });

            const blob = new Blob([response.data], { type: response.headers['content-type'] });
            const url = window.URL.createObjectURL(blob);
            const a = document.createElement('a');
            a.href = url;
            a.download = `contacts_${new Date().toISOString()}.csv`;
            a.click();
            window.URL.revokeObjectURL(url);
        } catch (error) {
            console.error("Error exporting contacts:", error);
            throw handleApiError(error);
        }
    }
};

export default contactService;

File: server/app/services/freshdesk/contact_service.rb
module Freshdesk
    class ContactService
        def initialize
            @client = ApiClient.new
        end

        def list_contacts(params = {})
            begin
                default_params = {
                    order_by: 'created_at',
                    order_type: 'desc'
                }

                query = default_params.merge(params)
                @client.request(:get, '/contacts', {}, query)
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to list contacts: #{e.message}")
            end
        end

        def get_contact(id)
            begin
                @client.request(:get, "/contacts/#{id}")
            rescue Freshdesk::ResourceNotFoundError
                raise Freshdesk::ResourceNotFoundError.new("Contact ##{id} not found")
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to get contact ##{id}: #{e.message}")
            end
        end

        def create_contact(contact_params, avatar = nil)
            if contact_params["tags"].is_a?(String) && contact_params["tags"].start_with?("[")
                contact_params["tags"] = JSON.parse(contact_params["tags"])
            end

            if contact_params["other_emails"].is_a?(String) && contact_params["other_emails"].start_with?("[")
                contact_params["other_emails"] = JSON.parse(contact_params["other_emails"])
            end

            if contact_params["company_id"].is_a?(String)
                contact_params["company_id"] = contact_params["company_id"].to_i
            end

            begin
                if avatar
                    @client.request_with_attachments('/contacts', contact_params, [avatar], :post)
                else
                    @client.request(:post, '/contacts', contact_params)
                end
            rescue Freshdesk::ValidationError => e
                raise e
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to create contact: #{e.message}")
            end
        end

        def update_contact(id, contact_params, avatar = nil)

            if contact_params["tags"].is_a?(String) && contact_params["tags"].start_with?("[")
                contact_params["tags"] = JSON.parse(contact_params["tags"])
            end

            if contact_params["other_emails"].is_a?(String) && contact_params["other_emails"].start_with?("[")
                contact_params["other_emails"] = JSON.parse(contact_params["other_emails"])
            end

            if contact_params["company_id"].is_a?(String)
                contact_params["company_id"] = contact_params["company_id"].to_i
            end

            begin
                if avatar
                    @client.request_with_attachments("/contacts/#{id}", contact_params, [avatar], :put)
                else
                    @client.request(:put, "/contacts/#{id}", contact_params)
                end
            rescue Freshdesk::ResourceNotFoundError
                raise Freshdesk::ResourceNotFoundError.new("Contact ##{id} not found")
            rescue Freshdesk::ValidationError => e
                raise e
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to update contact ##{id}: #{e.message}")
            end
        end

        def delete_contact(id)
            begin
                @client.request(:delete, "/contacts/#{id}")
            rescue Freshdesk::ResourceNotFoundError
                raise Freshdesk::ResourceNotFoundError.new("Contact ##{id} not found")
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to delete contact ##{id}: #{e.message}")
            end
        end

        def merge_contacts(primary_contact_id, secondary_contact_ids, contact_data = nil)
            begin
                primary_contact_id = primary_contact_id.to_i
                secondary_contact_ids = [secondary_contact_ids].flatten.map(&:to_i)

                endpoint = "/contacts/merge"
                merge_data = {
                    primary_contact_id: primary_contact_id,
                    secondary_contact_ids: secondary_contact_ids
                }

                if contact_data.present?
                    if contact_data[:other_emails].present? && contact_data[:other_emails].is_a?(ActionController::Parameters)
                        contact_data[:other_emails] = contact_data[:other_emails].to_unsafe_h.values
                    end

                    if contact_data[:company_ids].present? && contact_data[:company_ids].is_a?(ActionController::Parameters)
                        contact_data[:company_ids] = contact_data[:company_ids].to_unsafe_h.values
                    end

                    merge_data[:contact] = contact_data
                end

                @client.request(:post, endpoint, merge_data)
            rescue Freshdesk::ResourceNotFoundError
                raise Freshdesk::ResourceNotFoundError.new("Primary contact ##{primary_contact_id} not found")
            rescue Freshdesk::ValidationError => e
                raise e
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to merge contacts: #{e.message}")
            end
        end

        def get_contact_fields
            begin
                @client.request(:get, '/contact_fields')
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to get contact fields: #{e.message}")
            end
        end

        def get_companies
            begin
                @client.request(:get, '/companies')
            rescue Freshdesk::Error => e
                raise e
            rescue => e
                raise Freshdesk::RequestError.new("Failed to get companies: #{e.message}")
            end
        end
    end
end
